File: tests/clients/battery.go
// The Go half of the cross-client conformance battery.
//
// Runs the same scenarios as battery.mjs and battery.py against the same peer.
// This is the only driver in a different language runtime, so it has its own
// JSON encoder, HTTP client and SSE parser: every assumption shared between a
// Node gateway and a Node client is absent here.
//
// Usage: battery <url> <http|sse>
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strconv"
	"time"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

const (
	separators = "before\u2028middle\u2029after"
	unicodeStr = "a👩‍👩‍👧‍👦b عربى c éèê d 漢字 e 𝔘𝔫𝔦 f"
	plain      = "plain-result"
)

type result struct {
	Name   string `json:"name"`
	OK     bool   `json:"ok"`
	Detail string `json:"detail"`
}

var results []result

func record(name string, ok bool, detail string) {
	results = append(results, result{Name: name, OK: ok, Detail: detail})
}

func textOf(res *mcp.CallToolResult) string {
	if res == nil {
		return ""
	}
	for _, c := range res.Content {
		if tc, ok := c.(*mcp.TextContent); ok {
			return tc.Text
		}
	}
	return ""
}

func clip(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

func main() {
	url, kind := os.Args[1], os.Args[2]
	bigLength := 1024 * 1024
	if v := os.Getenv("BIG_LENGTH"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			bigLength = n
		}
	}

	defer func() {
		out, _ := json.Marshal(results)
		fmt.Println(string(out))
	}()

	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()

	client := mcp.NewClient(&mcp.Implementation{Name: "battery-go", Version: "1.0.0"}, nil)
	var transport mcp.Transport
	if kind == "sse" {
		transport = mcp.NewSSEClientTransport(url, nil)
	} else {
		transport = mcp.NewStreamableClientTransport(url, nil)
	}

	session, err := client.Connect(ctx, transport)
	if err != nil {
		record("connect", false, clip(err.Error(), 100))
		return
	}
	defer session.Close()
	record("connect", true, "")

	// tools/list
	if tools, err := session.ListTools(ctx, nil); err != nil {
		record("tools/list", false, clip(err.Error(), 60))
	} else {
		var names []string
		for _, t := range tools.Tools {
			names = append(names, t.Name)
		}
		sort.Strings(names)
		want := []string{"echo", "large", "plain", "separators", "shadowedError",
			"slow", "throws", "toolError", "unicode"}
		got, _ := json.Marshal(names)
		wantJSON, _ := json.Marshal(want)
		record("tools/list", string(got) == string(wantJSON), string(got))
	}

	simple := []struct{ label, tool, expect string }{
		{"plain text", "plain", plain},
		{"line separators", "separators", separators},
		{"unicode", "unicode", unicodeStr},
	}
	for _, c := range simple {
		res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: c.tool})
		if err != nil {
			record(c.label, false, clip(err.Error(), 60))
			continue
		}
		got := textOf(res)
		record(c.label, got == c.expect, clip(strconv.Quote(got), 60))
	}

	if res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: "large"}); err != nil {
		record(fmt.Sprintf("1 MB result"), false, clip(err.Error(), 70))
	} else {
		got := textOf(res)
		record("1 MB result", len(got) == bigLength, fmt.Sprintf("length %d", len(got)))
	}

	if res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: "toolError"}); err != nil {
		record("tool error stays a result", false, "threw: "+clip(err.Error(), 50))
	} else {
		record("tool error stays a result", res.IsError, fmt.Sprintf("isError=%v", res.IsError))
	}

	if res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: "throws"}); err != nil {
		record("throwing tool reported", true, "rejected: "+clip(err.Error(), 40))
	} else {
		record("throwing tool reported", res.IsError || textOf(res) != "",
			fmt.Sprintf("isError=%v", res.IsError))
	}

	args := map[string]any{
		"text":   unicodeStr,
		"number": -0.5,
		"flag":   false,
		"nested": map[string]any{"a": []any{1, nil, true}, "b": map[string]any{"c": separators}},
	}
	if res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: "echo", Arguments: args}); err != nil {
		record("arguments round trip", false, clip(err.Error(), 60))
	} else {
		var got map[string]any
		_ = json.Unmarshal([]byte(textOf(res)), &got)
		wantJSON, _ := json.Marshal(args)
		gotJSON, _ := json.Marshal(got)
		record("arguments round trip", string(gotJSON) == string(wantJSON), clip(string(gotJSON), 70))
	}

	started := time.Now()
	if res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: "slow"}); err != nil {
		record("slow call completes", false, clip(err.Error(), 60))
	} else {
		record("slow call completes", textOf(res) == "slow-done",
			fmt.Sprintf("%dms", time.Since(started).Milliseconds()))
	}

	if res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: "shadowedError"}); err != nil {
		record("result with an error field", false, clip(err.Error(), 60))
	} else {
		record("result with an error field", textOf(res) == "ok", strconv.Quote(textOf(res)))
	}

	// five concurrent calls
	type outcome struct {
		text string
		err  error
	}
	ch := make(chan outcome, 5)
	for i := 0; i < 5; i++ {
		go func() {
			res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: "plain"})
			ch <- outcome{textOf(res), err}
		}()
	}
	allOK := true
	for i := 0; i < 5; i++ {
		o := <-ch
		if o.err != nil || o.text != plain {
			allOK = false
		}
	}
	record("5 concurrent calls", allOK, "5 replies")

	if res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: "noSuchTool"}); err != nil {
		record("unknown tool errors", true, "rejected: "+clip(err.Error(), 40))
	} else {
		record("unknown tool errors", res.IsError, fmt.Sprintf("isError=%v", res.IsError))
	}
}

File: tests/clients/battery.mjs
// The Node half of the cross-client conformance battery.
//
// Runs the same scenarios the other language drivers run, so the results line up
// into a matrix. This one is the baseline: it uses the same SDK the gateway
// does, so anything failing here is the gateway, while anything failing only in
// another client is a disagreement between implementations.
const sdk = process.env.BATTERY_SDK ?? '@modelcontextprotocol/sdk'
const { Client } = await import(`${sdk}/client/index.js`)
const { SSEClientTransport } = await import(`${sdk}/client/sse.js`)
const StreamableHTTPClientTransport =
  process.argv[3] === 'sse'
    ? undefined
    : (await import(`${sdk}/client/streamableHttp.js`))
        .StreamableHTTPClientTransport

const URL_ = process.argv[2]
const KIND = process.argv[3]

const SEPARATORS = 'before' + '\u2028' + 'middle' + '\u2029' + 'after'
const UNICODE = 'a👩‍👩‍👧‍👦b عربى c éèê d 漢字 e 𝔘𝔫𝔦 f'
const PLAIN = 'plain-result'
const BIG_LENGTH = Number(process.env.BIG_LENGTH ?? 1024 * 1024)

const results = []
const record = (name, ok, detail = '') => results.push({ name, ok, detail })

const text = (res) => res?.content?.find((c) => c.type === 'text')?.text

const client = new Client(
  { name: 'battery-node', version: '1.0.0' },
  { capabilities: {} },
)
const transport =
  KIND === 'sse'
    ? new SSEClientTransport(new URL(URL_))
    : new StreamableHTTPClientTransport(new URL(URL_))

try {
  await client.connect(transport)
  record('connect', true, client.getServerVersion?.()?.name ?? '')
} catch (e) {
  record('connect', false, String(e.message).slice(0, 80))
  console.log(JSON.stringify(results))
  process.exit(0)
}

const call = async (name, args = {}) =>
  client.callTool({ name, arguments: args })

try {
  const tools = (await client.listTools()).tools.map((t) => t.name).sort()
  const want = [
    'echo',
    'large',
    'plain',
    'separators',
    'shadowedError',
    'slow',
    'throws',
    'toolError',
    'unicode',
  ]
  record(
    'tools/list',
    JSON.stringify(tools) === JSON.stringify(want),
    tools.join(),
  )
} catch (e) {
  record('tools/list', false, String(e.message).slice(0, 60))
}

for (const [name, tool, expect] of [
  ['plain text', 'plain', PLAIN],
  ['line separators', 'separators', SEPARATORS],
  ['unicode', 'unicode', UNICODE],
]) {
  try {
    const got = text(await call(tool))
    record(
      name,
      got === expect,
      got === expect ? '' : `got ${JSON.stringify(got).slice(0, 60)}`,
    )
  } catch (e) {
    record(name, false, String(e.message).slice(0, 60))
  }
}

try {
  const got = text(await call('large'))
  record('1 MB result', got?.length === BIG_LENGTH, `length ${got?.length}`)
} catch (e) {
  record('1 MB result', false, String(e.message).slice(0, 60))
}

try {
  const res = await call('toolError')
  record(
    'tool error stays a result',
    res?.isError === true,
    `isError=${res?.isError}`,
  )
} catch (e) {
  record(
    'tool error stays a result',
    false,
    `threw: ${String(e.message).slice(0, 50)}`,
  )
}

try {
  const res = await call('throws')
  // The SDK surfaces a thrown tool as an isError result rather than rejecting.
  record(
    'throwing tool reported',
    res?.isError === true || Boolean(text(res)),
    `isError=${res?.isError}`,
  )
} catch (e) {
  record(
    'throwing tool reported',
    true,
    `rejected: ${String(e.message).slice(0, 40)}`,
  )
}

try {
  const args = {
    text: UNICODE,
    number: -0.5,
    flag: false,
    nested: { a: [1, null, true], b: { c: SEPARATORS } },
  }
  const got = JSON.parse(text(await call('echo', args)))
  record(
    'arguments round trip',
    JSON.stringify(got) === JSON.stringify(args),
    JSON.stringify(got).slice(0, 70),
  )
} catch (e) {
  record('arguments round trip', false, String(e.message).slice(0, 60))
}

try {
  const started = Date.now()
  const got = text(await call('slow'))
  record(
    'slow call completes',
    got === 'slow-done',
    `${Date.now() - started}ms`,
  )
} catch (e) {
  record('slow call completes', false, String(e.message).slice(0, 60))
}

try {
  const res = await call('shadowedError')
  record(
    'result with an error field',
    text(res) === 'ok',
    `text=${JSON.stringify(text(res))}`,
  )
} catch (e) {
  record('result with an error field', false, String(e.message).slice(0, 60))
}

try {
  const got = await Promise.all(Array.from({ length: 5 }, () => call('plain')))
  record(
    '5 concurrent calls',
    got.every((r) => text(r) === PLAIN),
    `${got.length} replies`,
  )
} catch (e) {
  record('5 concurrent calls', false, String(e.message).slice(0, 60))
}

try {
  const res = await call('noSuchTool')
  record(
    'unknown tool errors',
    Boolean(res?.isError),
    `isError=${res?.isError}`,
  )
} catch (e) {
  record(
    'unknown tool errors',
    true,
    `rejected: ${String(e.message).slice(0, 40)}`,
  )
}

await client.close().catch(() => {})
console.log(JSON.stringify(results))
